Fixes #30241 - Fix react eslint tests

johnpmitsch opened the pull request.

jeremylenz commented on webpack/.eslintrc.js:
```diff
@@ -50,6 +50,11 @@ module.exports = {
     'promise/prefer-await-to-callbacks': 'error',
+    "react/forbid-prop-types": [
+      'error',
+      // Obmitting "array" as older codebase needs to be updated to use this by default
```
```suggestion
      // Omitting "array" as older codebase needs to be updated to use this by default
```

johnpmitsch replied:
:flushed:

jeremylenz commented on webpack/scenes/Subscriptions/components/SubscriptionsTable/components/Dialogs/index.js:
```diff
@@ -17,10 +17,10 @@ const Dialogs = ({
-  unsavedChangesDialog: PropTypes.shape(UnsavedChangesDialog.propTypes).isRequired,
-  inputsErrorsDialog: PropTypes.shape(InputsErrorsDialog.propTypes).isRequired,
-  deleteDialog: PropTypes.shape(DeleteDialog.propTypes).isRequired,
+  updateDialog: PropTypes.shape({}).isRequired,
+  unsavedChangesDialog: PropTypes.shape({}).isRequired,
+  inputsErrorsDialog: PropTypes.shape({}).isRequired,
+  deleteDialog: PropTypes.shape({}).isRequired,
```
Are these necessary?  I would think they would all return objects so should be okay.

johnpmitsch replied:
This was failing tests as well

jeremylenz replied:
:man_shrugging:

johnpmitsch commented:
Looking further it seems to be `eslint-plugin-react` update, even though it a patch version: 7.20.0 -> 7.20.1

Pinning to 7.20.0, I don't see the error, but with 7.20.1 installed, I do
```diff
diff --git a/package.json b/package.json
index 21cc5dc..150439e 100644
--- a/package.json
+++ b/package.json
@@ -37,7 +37,7 @@
     "eslint-plugin-jest": "^21.18.0",
     "eslint-plugin-jsx-a11y": "^6.0.2",
     "eslint-plugin-promise": "^4.2.1",
-    "eslint-plugin-react": "^7.4.0",
+    "eslint-plugin-react": "7.20.0",
     "jest": "^24.9.0",
     "nock": "^12.0.3",
     "react-test-renderer": "^16.0.0"
```

[here](https://github.com/yannickcr/eslint-plugin-react/commit/997a99f389a7df9673de890d71a17e2b6838f72b) are the changes between, I haven't dug through them specifically, but I see commits dealing with `forbid-prop-types` and the parser, which is where we saw issues.

johnpmitsch commented:
@jeremylenz The alternative fix for the array proptype errors is to add eslint-ignore comments to the existing lines that use `PropType.array` in our existing codebase. There were maybe 10 of them iirc. 

This way, we won't be able to use `PropTypes.array` in the future, instead using `PropTypes.arrayOf`, specifying the type of value in the array as well.

I can update to that if you are on board as well, if we need the fix sooner we can merge this and discuss a follow-up change.

johnpmitsch commented:
> @jeremylenz The alternative fix for the array proptype errors is to add eslint-ignore comments to the existing lines that use `PropType.array` in our existing codebase. There were maybe 10 of them iirc.
> 
> This way, we won't be able to use `PropTypes.array` in the future, instead using `PropTypes.arrayOf`, specifying the type of value in the array as well.
> 
> I can update to that if you are on board as well, if we need the fix sooner we can merge this and discuss a follow-up change.

Here are the errors after removing the eslint rule override: (I guess it's 23 of them :smile:)

```
[vagrant@centos7 katello]$ npm run lint

> katello@1.0.0 lint /home/vagrant/katello
> npx eslint webpack/


/home/vagrant/katello/webpack/components/Content/ContentTable.js
  37:5  error  Prop type `array` is forbidden  react/forbid-prop-types

/home/vagrant/katello/webpack/components/Content/Details/ContentDetails.js
  62:5  error  Prop type `array` is forbidden  react/forbid-prop-types
  63:5  error  Prop type `array` is forbidden  react/forbid-prop-types
  64:5  error  Prop type `array` is forbidden  react/forbid-prop-types

/home/vagrant/katello/webpack/scenes/AnsibleCollections/Details/AnsibleCollectionDetails.js
  86:5  error  Prop type `array` is forbidden  react/forbid-prop-types
  87:5  error  Prop type `array` is forbidden  react/forbid-prop-types

/home/vagrant/katello/webpack/scenes/ModuleStreams/Details/ModuleStreamDetails.js
  84:5  error  Prop type `array` is forbidden  react/forbid-prop-types
  85:5  error  Prop type `array` is forbidden  react/forbid-prop-types
  86:5  error  Prop type `array` is forbidden  react/forbid-prop-types

/home/vagrant/katello/webpack/scenes/RedHatRepositories/RedHatRepositoriesPage.js
  114:5  error  Prop type `array` is forbidden  react/forbid-prop-types

/home/vagrant/katello/webpack/scenes/RedHatRepositories/components/EnabledRepository/EnabledRepository.js
  97:5  error  Prop type `array` is forbidden  react/forbid-prop-types

/home/vagrant/katello/webpack/scenes/RedHatRepositories/components/RepositorySetRepository/RepositorySetRepository.js
  144:5  error  Prop type `array` is forbidden  react/forbid-prop-types

/home/vagrant/katello/webpack/scenes/Subscriptions/Details/SubscriptionDetailAssociations.js
  48:5  error  Prop type `array` is forbidden  react/forbid-prop-types

/home/vagrant/katello/webpack/scenes/Subscriptions/Details/SubscriptionDetailProductContent.js
  53:5  error  Prop type `array` is forbidden  react/forbid-prop-types

/home/vagrant/katello/webpack/scenes/Subscriptions/Details/SubscriptionDetailProducts.js
  21:5  error  Prop type `array` is forbidden  react/forbid-prop-types

/home/vagrant/katello/webpack/scenes/Subscriptions/Manifest/ManageManifestModal.js
  355:5  error  Prop type `array` is forbidden  react/forbid-prop-types

/home/vagrant/katello/webpack/scenes/Subscriptions/SubscriptionsPage.js
  318:5  error  Prop type `array` is forbidden  react/forbid-prop-types
  319:5  error  Prop type `array` is forbidden  react/forbid-prop-types
  320:5  error  Prop type `array` is forbidden  react/forbid-prop-types
  321:5  error  Prop type `array` is forbidden  react/forbid-prop-types

/home/vagrant/katello/webpack/scenes/Subscriptions/UpstreamSubscriptions/UpstreamSubscriptionsPage.js
  265:5  error  Prop type `array` is forbidden  react/forbid-prop-types

/home/vagrant/katello/webpack/scenes/Subscriptions/components/SubscriptionsTable/SubscriptionsTable.js
  288:5  error  Prop type `array` is forbidden  react/forbid-prop-types

/home/vagrant/katello/webpack/scenes/Subscriptions/components/SubscriptionsTable/components/Table.js
  102:5  error  Prop type `array` is forbidden  react/forbid-prop-types

```

jeremylenz commented:
I'm ok with adding the eslint ignore comments for now.  (If any of them are super obvious, may as well just change them to `PropTypes.arrayOf` now rather than adding to the technical debt?  Up to you..)

johnpmitsch commented:
> I'm ok with adding the eslint ignore comments for now. (If any of them are super obvious, may as well just change them to `PropTypes.arrayOf` now rather than adding to the technical debt? Up to you..)

Sounds good, this way the rule will work as intended in the future rather than partially disabled in the eslint config. 

I just added the disable comments to the breakages, I also added a note on each one just so it's clear that we don't want to disable this rule any further and this was a special circumstance.

johnpmitsch commented:
[test katello]

johnpmitsch commented:
[test katello]

johnpmitsch commented:
[test katello]
